### PR TITLE
JitArm64: Use LogicalImm in boolX

### PR DESCRIPTION
**Pull request by Sintendo**
ARM64 has a special logical immediate encoding scheme, that can be used with `AND`, `ORR`, and `EOR`. By taking advantage of this, we no longer need to materialize the immediate value in a register, saving instructions and/or reducing register pressure.

---

<details><summary>andx</summary>

Before:
```
mov    w26, #-0x80000000
and    w27, w27, w26
sxtw   x24, w27
```

After:
```
and    w27, w27, #0x80000000
sxtw   x26, w27
```
</details>

<details><summary>orx</summary>

Before:
```
mov    w23, #0x1
orr    w23, w25, w23
```

After:
```
orr    w23, w25, #0x1
```
</details>

<details><summary>norx</summary>

Before:
```
mov    w25, #-0x2001
orr    w23, w23, w25
mvn    w23, w23
```

After:
```
orr    w23, w23, #0xffffdfff
mvn    w23, w23
```
</details>

<details><summary>xorx</summary>

Before:
```
mov    w23, #0x1e
eor    w23, w27, w23
```

After:
```
eor    w23, w27, #0x1e
```
</details>

<details><summary>eqvx</summary>

Before:
```
mov    w23, #0x4
eon    w26, w23, w22
```

After:
```
eor    w26, w22, #0xfffffffb
```
</details>

---

This one has been removed.

<s><details><summary>andcx</summary>

Before:
```
mov    w24, #-0x20
bic    w27, w24, w26
```

After:
```
mvn    w27, w26
and    w27, w27, #0xffffffe0
```
</details></s>

### CODE REVIEW

**Comment by JosJuice** [comment]
Overall this is a good idea, but your example output for andcx is worse than before. While the instruction count is the same as before, the critical path becomes one cycle longer.

**Comment by JosJuice** [comment]
Now that I think about it, even if both the cycle counts and the critical path latency were the same, the old one that materializes the immediate value would be preferable. This is because if the register later has to be written back to the ppcState struct, it has to be materialized sooner or later anyway (unless it's 0). So: If you can save one instruction by not materializing the immediate, please do so (it helps in the cases where the register doesn't have to be written to ppcState), but otherwise you should keep materializing the immediate.

**Comment by Sintendo** [comment]
That's an excellent point. Indeed, in cases where the instruction sequence is equal in length, materializing the immediate in a register could allow subsequent uses to leech off of it. Effectively, this means we shouldn't do the optimization for `andcx` and `orcx` (except in cases where operand 2 is the immediate and we can precompute its complement). I have updated the PR accordingly.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:418]: In the `final_not` case, we could make use of de Morgan's laws and turn ~(s & b) into ~s | ~b. Since inverting the immediate has no runtime cost, this would let us replace the AND+MVN with ORN. But inverting the immediate after we already have `log_imm` seems like effort... So I'll leave it up to you if you want to try implementing this in this PR or not.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:445]: If this is a bug, it must be a pre-existing bug, but... This shouldn't run for `complement_b`, should it? If we have `is_zero && complement_b`, that means `s` is an immediate zero, so the result should be zero regardless of what `b` is. (If the immediate zero had been in `b`, code earlier would have set `complement_b` to false.)

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:462]: Same here regarding de Morgan's laws.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp
Comment (Sintendo) [reply Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:445]: We're doing `is_or`. There's three possibilities.

1. `final_not` with immediate in s: `~(0 | b) == ~b == ~j`
2. `final_not` with immediate in b: `~(s | 0) == ~s == ~j`
3. `complement_b` (immediate can only be in s): `0 | ~b == ~b == ~j`

In all cases, it works out to `~j`. This seems correct to me, but perhaps I'm missing something.

Comment (JosJuice) [reply Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:445]: Oh, right, it's OR and not AND. Yeah, ignore me then.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp
Comment (Sintendo) [reply Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:418]: Sorry, forgot about this for a while.

Not a bad idea. You can even use this approach for any immediate, not just those that can be expressed as LogicalImm. But you still need to materialize the immediate somehow and that might take more than one MOV instruction, in which case using LogicalImm might still be preferable...

I should also note that I haven't seen a single game use `nand` with immediates. And the only game that I've seen use `nor` is Zelda Master Quest.

So given the complexity and how uncommon these instruction patterns are, I think it would be better to leave this for a follow-up PR for now, if that's alright with you.

**Review by JosJuice — APPROVED** [review]
I'm planning to merge this after the beta is out.

**Pull request merged**
